### PR TITLE
fix: change random.word to be based on definitions.word.*

### LINKED ISSUE #113: `random.word()` can return undesirable non-alpha characters
### Describe the bug

The pool of strings from which `random.words()` pulls values includes strings that have parentheses, e.g. `Cocos (Keeling) Islands` and `Killer Whale (Orca)`. This can result in `(Keeling)` and `(Orca)` being returned.

There's a larger problem here of non-alphabetic characters being returned in words. Obviously some non-alpha characters are desired in words, e.g. `-`, but some are not. We should decide on the set of acceptable characters that `word()` can return and regex the others out.

### Reproduction

Mock `faker.fake` with the above mentioned values and you can see the bad returns.

### Additional Info

### PR DESCRIPTION
**Pull request by pkuczynski**
Fixes #113
Fixes #218

### CODE REVIEW

**Comment by prisis** [comment]
I would put into 6.1  @import-brain

**Comment by import-brain** [comment]
> I would put into 6.1  @import-brain

Done

**Comment by pkuczynski** [comment]
How about the review @prisis ?

**Comment by prisis** [comment]
@pkuczynski hope it's fine with you if i wait with the review a bit.  
  
I would like to fix the tests more and remove all the mock function call, so we now that we break something.  
  
And this is a runtime change of the code, so we will not merge it into v6.0 for now. But im thankful for you spend time on this PR.

File: src/random.ts [hunk 1]
```diff
@@ -159,53 +159,28 @@ export class Random {
     return this.faker.datatype.boolean();
   }
 
-  // TODO: have ability to return specific type of word? As in: noun, adjective, verb, etc
```
Comment (prisis) [inline src/random.ts:162]: The question is if we want to have this feature too.

**Review by prisis — APPROVED** [review]

File: src/random.ts (hunk 1 above)
Comment (pkuczynski) [reply src/random.ts:162]: We already have. Users can use `faker.word.xyz` to do it... That's why I removed this comment.

**Comment by pkuczynski** [comment]
The problem is, that the longer you wait with merging, the higher chance of conflict. I worked on this as it was listed as "help needed". This feature does not have any tests, so not sure what mocks do you want to remove?

**Comment by Shinigami92** [comment]
> The problem is, that the longer you wait with merging, the higher chance of conflict. I worked on this as it was listed as "help needed". This feature does not have any tests, so not sure what mocks do you want to remove?

Sorry but this is irrelevant to the current situation of the project.

We have a roadmap https://github.com/orgs/faker-js/projects/2 with the milestone https://github.com/faker-js/faker/milestone/1
We are thanksful that you try to contribute, but if you want to prevent merge conflicts right now, than wait until we release 6.0.0 and stop opening PR for now

This said, I can somewhat promise you that `src/random.ts` will not change that much anymore. The project stability is more regarding to the test-suite, bundling-setup and documentation. Multiple times said: no runtime changes before 6.0.0 released

File: src/random.ts [hunk 2]
```diff
@@ -159,53 +159,28 @@ export class Random {
...
-    const result = this.faker.fake('{{' + randomWordMethod + '}}');
-    return this.faker.random.arrayElement(result.split(' '));
+    const definition = this.faker.random.arrayElement(definitions);
+    let word = '';
+
+    do {
+      word = this.faker.random.arrayElement(definition);
+    } while (word.includes(' '));
+
+    return word;
```
Comment (griest024) [inline src/random.ts:183]: I much prefer the `split` implementation over the `do...while`.

**Review by griest024 — CHANGES_REQUESTED** [review]

File: src/random.ts (hunk 2 above)
Comment (pkuczynski) [reply src/random.ts:183]: The `do...while` is only used here to get next random word if the one currently selected contains a space. Not sure how split would help here? Can you show an example?

Comment (griest024) [reply src/random.ts:183]: For example if `contact lens` was selected, `do..while` would skip that whereas `split` would return either `contact` or `lens`. This is assuming that we need to eliminate spaces (which maybe we should not?).

Comment (griest024) [reply src/random.ts:183]: I vote remove this logic entirely and remove word examples with spaces instead. A regex search for `'.* .*'` in `src/locales/*/word` should get you all of them.

Comment (pkuczynski) [reply src/random.ts:183]: Sure, let's do that! I will remove this logic here...

**Review by griest024 — APPROVED** [review]

**Review by ST-DDT — APPROVED** [review]

**Comment by pkuczynski** [comment]
Possibly in the future we might consider to move this to faker.word.any() as this what happened with other methods from this module...

**Comment by griest024** [comment]
> Possibly in the future we might consider to move this to faker.word.any() as this what happened with other methods from this module...
> 
> What do you think? I could record this idea as a GitHub issue for future consideration...

I think that is reasonable.

**Comment by pkuczynski** [comment]
> I think that is reasonable.

Here you go https://github.com/faker-js/faker/issues/339 :)

**Comment by Shinigami92** [comment]
@pkuczynski Could you rebase this PR so we can see how it behaves with all our new tests?

**Comment by Shinigami92** [comment]
As this is changing runtime behavior and we already want to do #339 anyways, I will close this PR in favor of #339

**Comment by pkuczynski** [comment]
@Shinigami92 I think this should be still merged, because it contains valid changes. Then the move in #339 can happen more easily...

**Comment by Shinigami92** [comment]
@ST-DDT and I decided this together via screen-share some minutes ago
We do not see to merge this because it just changes behavior that we already want to deprecate anyways.
What you could do right now would be to just add a `console.warn` that logs that this function will be deprecated so we can already deprecate it in `v6.1`.

Let me also say that I do totally **not** expect that the move from initial to v6.0.0 will take same amount of time as from v6.0 to v6.1 🙂 
We want now to move on a bit faster

**Comment by pkuczynski** [comment]
I mean this function after the changes done currently, can be with pretty much (almost) no change moved to `word.any` and `random.word` deprecated. It does slightly change the behaviour, that's true but not in a way which is dangerous in any way and it already does fix the issues reported in #113... So I still see this as perfectly mergeable...

**Comment by pkuczynski** [comment]
@Shinigami92 @ST-DDT rebased and tests fixed

**Review by Shinigami92 — CHANGES_REQUESTED** [review]
I'm more and more tempted to not accept this PR 😕 Sorry...

I re-read through #113 and I thing we should go with the suggested proposal to just regex them out if they generate _"invalid"_ words and then make a retry

With this PR (#276) we loose many possible words, especially hacker words, were you can see that they are not in generated system filenames anymore

I will block this from my side with a `request changes`, but if another maintainer is on my side, then I would vote for close this PR

**Comment by pkuczynski** [comment]
Well, at least we get more sensible words with this pr comparing to before... but yeah, do what you wish...

**Comment by ST-DDT** [comment]
> Well, at least we get more sensible words with this pr comparing to before... but yeah, do what you wish...

What do you consider a "sensible" word? IMO we have to decide first what we expect from this method.
AFAIK this method is used in file name generation. Should those contain Chinese letters in Chinese locales?

**Comment by pkuczynski** [comment]
> > Well, at least we get more sensible words with this pr comparing to before... but yeah, do what you wish...
> 
> What do you consider a "sensible" word? IMO we have to decide first what we expect from this method. AFAIK this method is used in file name generation. Should those contain Chinese letters/words in Chinese locales? What about the opposite? Should it partially return English file names in Chinese locales

I wold consider noun, adjective, verb, ... a sensible word. While things like `streetSuffix` or `currencyName` does not seem something I would expect from `random.word` (or `word.any` in the future). For example would `as long as` returned at the moment could be considered as a word? Even if you split it into `as` and `long`. The second one seems ok, but `as`? Don't think so...

Looking at the wikipedia https://en.wikipedia.org/wiki/Word:

> In [linguistics](https://en.wikipedia.org/wiki/Linguistics), a word of a [spoken language](https://en.wikipedia.org/wiki/Spoken_language) can be defined as the smallest sequence of [phonemes](https://en.wikipedia.org/wiki/Phoneme) that can be uttered in isolation with [objective](https://en.wikipedia.org/wiki/Semantics) or [practical](https://en.wikipedia.org/wiki/Pragmatics) [meaning](https://en.wikipedia.org/wiki/Meaning_(linguistics)). In many languages, words also correspond to sequences of [graphemes](https://en.wikipedia.org/wiki/Grapheme) ("letters") in their standard [writing systems](https://en.wikipedia.org/wiki/Writing_system) that are delimited by spaces wider than the normal inter-letter space, or by other graphical conventions.[[1]](https://en.wikipedia.org/wiki/Word#cite_note-hasp2011-1) The concept of "word" is usually distinguished from that of a [morpheme](https://en.wikipedia.org/wiki/Morpheme), which is the smallest unit of word which has a meaning, even if it will not stand on its own together or in other small words.

**Comment by pkuczynski** [comment]
For filename generation we could also use `lorem.slug()`.

In other languages, I would expect the filename is generated in the selected locale, unless there is no translation for it. Wouldn't you?

**Comment by pkuczynski** [comment]
I am not doing anything with his PR until you make a decision, right?

**Comment by Shinigami92** [comment]
In yesterdays maintainers-team-meeting we discussed that we want to have the `random.word` function just as it is right now, but add a `@deprecated` JSDoc comment to it and a warning for removal
We want to have the `word.any` function directly and it should be implemented like you are already doing in this PR

**Comment by pkuczynski** [comment]
OK!

**Pull request closed without merging**